### ISSUE TITLE
incorrect instruction size - vmovd

### ISSUE THREAD

**Issue by 2xic** [unicorn-engine/unicorn#1088]
Hey!

Unicorn seems to report an incorrect instruction size for the vmovd instruction. Here is basic script to replicate the issue.

```
from unicorn import *
from unicorn.x86_const import * 
from keystone import *

CODE = b"vmovd xmm0, esi;"
UNICORN_CODE = None
try:
	ks = Ks(KS_ARCH_X86, KS_MODE_64)
	encoding, count = ks.asm(CODE)
	print("keystone instruction size: %i" % (len(encoding)))
	UNICORN_CODE = bytes(encoding)
except KsError as e:
	print("ERROR: %s" %e)


ADDRESS = 0x1000000
def hook_code(mu, address, size, user_data):  
	print("unicorn instruction size %i" % (size))

try:
	mu = Uc(UC_ARCH_X86, UC_MODE_64)
	mu.hook_add(UC_HOOK_CODE, hook_code)
	# map 2MB memory for this emulation
	mu.mem_map(ADDRESS, 2 * 1024 * 1024)

	# write machine code to be emulated to memory
	mu.mem_write(ADDRESS, UNICORN_CODE)

	mu.emu_start(ADDRESS, ADDRESS + len(UNICORN_CODE))
except UcError as e:
	print("ERROR: %s" % e)

```
**Output**

```
keystone instruction size: 4
unicorn instruction size 3
```
Tested on Debian Linux 64 bit with newest unicorn from github.

**Comment by 2xic** [comment]
So I have tried to debug the problem. The problem seems to be regarding parsing the vex opcode. After doing the switch check under "Collect prefixes", the switch statement for opcodes points to what I'm pretty sure is the wrong opcode. In _reswitch_ the switch(b), will have the value **0x6e**, however **0x6e** is a opcode for string ops. Which is interacting with I/O ports (seems wrong given that vmovd is moving qword between registers, not interacting with I/O ports). However if you look at **0xc6**(last byte of the instruction), it points to _mov Ev, Iv_, I feel this is more correct. 

So "vmovd xmm0, esi;" encodes to `['0xc5', '0xf9', '0x6e', '0xc6']`. **0xc5** means that this is a 2-byte VEX, you can see it read in the vex data at the line with the following code 
`int vex3, vex2 = cpu_ldub_code(env, s->pc);` (pc will here point at **0xf9**). A few liens later ` s->pc++;` is executed. All of this seems okay. However since b == 0xc5 there is a check` if (b == 0xc5)`,  it will read the opcode as `b = cpu_ldub_code(env, s->pc++);`. This is what I believe is wrong. Since the value it reads out is **0x6e** and **0x6e** is as I said a opcode for string ops. 

Patching the original switch statement under "Collect prefixes" to make the switch check for opcode in _reswitch_ point to **0xc6** makes the emulation exit normally. Without this patch the pc will go out of bounds. Since after the function _disas_insn_ is done, pc will point at **0xc6**. I just added this line as a basic patch, `b = cpu_ldub_code(env, s->pc);` after `b = cpu_ldub_code(env, s->pc++);` under the check ` if (b == 0xc5)`. I'm sure this will break some other instruction support. I haven't tested the patch with other instructions, only with the test script above. I have probably done something wrong, I doubt the solution is this simple.  However, I hope this analysis give more insight to people more knowledgeable about this special instruction (and x86 instruction encoding) and can make creating a good patch easier.

**Comment by Waterman178** [comment]
This was fixed with 5bf6d77. #1073

**Comment by 2xic** [comment]
That is correct. Running a fresh git clone of master I don't experience the problem anymore, I guess something must have gone wrong when I tried to fetch the updates from master before opening  this issues... Thanks Waterman178!
